### PR TITLE
Fixed #28889 -- Use JS to prevent double submission of admin forms (disable buttons)

### PR DESCRIPTION
**Pull request by mgaligniana**
This is the "disable submit buttons" approach for [ticket](https://code.djangoproject.com/ticket/28889)

["You already submitted (alert)" approach](https://github.com/django/django/pull/15217)

We added three tests:
- Buttons are disabled after click save
- Submit and go back
- Double click won't create multiple objects

You can run them doing:
`./runtests.py admin_views.test_prevent_double_submission.SeleniumTests --selenium=chrome`
`./runtests.py admin_views.test_prevent_double_submission.SeleniumTests --selenium=firefox`

### CODE REVIEW

File: django/contrib/admin/static/admin/js/change_form.js [hunk 1]
```diff
@@ -2,8 +2,35 @@
 {
     const inputTags = ['BUTTON', 'INPUT', 'SELECT', 'TEXTAREA'];
     const modelName = document.getElementById('django-admin-form-add-constants').dataset.modelName;
+    const buttons = document.querySelectorAll('input[type=submit]');
+    const disableSubmitButtons = () => {
+        buttons.forEach(button => {
+            button.setAttribute('disabled', true);
```
Comment (adamchainz) [inline django/contrib/admin/static/admin/js/change_form.js:8]: According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute#parameters) one should set boolean attrs to the empty string:

```suggestion
            button.setAttribute('disabled', '');
```

File: django/contrib/admin/static/admin/js/change_form.js [hunk 2]
```diff
@@ -2,8 +2,35 @@
 {
     const inputTags = ['BUTTON', 'INPUT', 'SELECT', 'TEXTAREA'];
     const modelName = document.getElementById('django-admin-form-add-constants').dataset.modelName;
+    const buttons = document.querySelectorAll('input[type=submit]');
```
Comment (adamchainz) [inline django/contrib/admin/static/admin/js/change_form.js:5]: If the user extends the admin with extra forms somehow, this query selector will capture those forms’ submit buttons. But later the submit event listener is specific to the one form. I think it's better to do this later, under `if (modelName)`, and find only submit buttons in the form with `form.querySelectorAll(...)`

Comment (mgaligniana) [reply django/contrib/admin/static/admin/js/change_form.js]: Oh, good catch!

Done!

**Comment by mgaligniana** [comment]
@adamchainz Sorry to bother you! But is there anything I can do to push this forward?

**Comment by adamchainz** [comment]
I am kinda waiting for opinions from the others who reviewed the original PR. Perhaps try commenting there. Emails can get buried with the new year.

**Comment by adamchainz** [comment]
Btw looking again at this approach - it could make sense to re-enable the buttons after a timeout (3 seconds? 10 seconds?) to allow them to work even when the page fails to load / the user cancels navigation... But otherwise I think it's pretty solid.

**Comment by mgaligniana** [comment]
Thanks for the answer!

Regarding the timeout: I wonder what would happen if the request takes longer than the set timeout, i.e. the buttons would be enabled but the request is not finished yet.

After reading again the @thibaudcolas [comment](https://github.com/django/django/pull/15217#issuecomment-999658111) and [yours](https://github.com/django/django/pull/15217#issuecomment-999720926) in the [first approach](https://github.com/django/django/pull/15217) I would say don't manage cancel or network errors

**Comment by carltongibson** [comment]
@mgaligniana Thanks for the work here. I've commented on #15217 — which I think we should proceed with. I've just asked to check if there's a consensus there for that. 

I'll close this for now, assuming that — we can always reopen if the discussion leans to this approach. 

🏅

**Pull request closed without merging**